examples/amm/src/contract.rs

```diff
@@ -0,0 +1,342 @@
+
+                if input_token_idx > 1 || output_token_idx > 1 {
+                    return Err(AmmError::InvalidTokenIdx);
+                }
+
+                if input_token_idx == output_token_idx {
+                    return Err(AmmError::EqualTokensError);
+                }
+
+                let input_pool_balance = self.get_pool_balance(input_token_idx).await?;
+                let output_pool_balance = self.get_pool_balance(output_token_idx).await?;
+
+                let output_amount = amm::calculate_output_amount(
+                    input_amount,
+                    input_pool_balance,
+                    output_pool_balance,
+                )?;
+
+                self.receive_from_account(&owner, input_token_idx, input_amount)
+                    .await?;
+                self.send_to(&owner, output_token_idx, output_amount).await
+            }
+            OperationType::AddLiquidity {
+                owner,
+                token0_amount,
+                token1_amount,
+            } => {
+                if token0_amount == Amount::ZERO || token1_amount == Amount::ZERO {
+                    return Err(AmmError::NoZeroAmounts);
+                }
+
+                let balance0 = self.get_pool_balance(0).await?;
+                let balance1 = self.get_pool_balance(1).await?;
+
+                if balance0 > Amount::ZERO && balance1 > Amount::ZERO {
+                    let one_amount_bigint =
+                        BigUint::from_u128(<Amount as Into<u128>>::into(Amount::ONE))
+                            .expect("Couldn't generate Amount::ONE in bigint");
+
+                    let balance0_bigint =
+                        BigUint::from_u128(<Amount as Into<u128>>::into(balance0))
+                            .expect("Couldn't generate balance0 in bigint");
+                    let balance1_bigint =
+                        BigUint::from_u128(<Amount as Into<u128>>::into(balance1))
+                            .expect("Couldn't generate balance1 in bigint");
+                    let token0_amount_bigint =
+                        BigUint::from_u128(<Amount as Into<u128>>::into(token0_amount))
+                            .expect("Couldn't generate token0_amount in bigint");
+                    let token1_amount_bigint =
+                        BigUint::from_u128(<Amount as Into<u128>>::into(token1_amount))
+                            .expect("Couldn't generate token1_amount in bigint");
+
+                    // To always maintain the precision, whenever we divide,
+                    // we always multiply by Amount::ONE before dividing
+                    let old_ratio_bigint =
+                        (balance0_bigint * one_amount_bigint.clone()) / balance1_bigint;
+                    let new_ratio_bigint =
+                        (token0_amount_bigint * one_amount_bigint) / token1_amount_bigint;
+
+                    if balance1 > Amount::ZERO && old_ratio_bigint != new_ratio_bigint {
```

Comment:
We need to allow small discrepancies here. Otherwise if e.g. one of the balances is prime, doubling the liquidity is the smallest amount you can add.

(We probably shouldn't measure the discrepancy in terms of the difference of the ratios, though, and find a measure that treats both tokens the same. E.g. compare `balance0 / token0_amount` to `balance1 / token1_amount` instead.)